### PR TITLE
nixos/redis: misc module changes

### PR DESCRIPTION
**Pull request by ajs124**
###### Things done
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ju1m** [comment]
@ajs124 thanks for that PR. Would it work the same by removing the `sed` logic and having `/var/lib/${redisName name}/redis.conf` just contain an `include /run/${redisName name}/nixos.conf` which will then include `redisConfStore` and `requirepass` if needed?

**Comment by ajs124** [comment]
That could probably work :thinking: 

I'll give it a try.

**Comment by ajs124** [comment]
tbh, redis might be writing too much stuff to this config file that people sometimes don't want to persist.
this is _necessary_ for some usecases, but maybe it breaks others?

depending on your opinion on the matter, I'd be open to make this optional.

File: nixos/modules/services/databases/redis.nix [hunk 1]
```diff
@@ -347,16 +347,24 @@ in {
...
-            {
-              printf requirePass' '
-              cat ${escapeShellArg conf.requirePassFile}
-            } >>/run/${redisName name}/redis.conf
-          '')
-        )];
+        ExecStart = "${cfg.package}/bin/redis-server /var/lib/${redisName name}/redis.conf ${escapeShellArgs conf.extraParams}";
+        ExecStartPre = "+"+pkgs.writeShellScript "${redisName name}-prep-conf" (let
+          redisConfVar = "/var/lib/${redisName name}/redis.conf";
+          redisConfRun = "/run/${redisName name}/pass.conf";
```
Comment (ju1m) [inline nixos/modules/services/databases/redis.nix:353]: Calling it `nixos.conf` would be more appropriate.

File: nixos/modules/services/databases/redis.nix [hunk 2]
```diff
@@ -347,16 +347,24 @@ in {
...
+        in ''
+          touch "${redisConfVar}" "${redisConfRun}"
+          chown '${conf.user}' "${redisConfVar}" "${redisConfRun}"
+          chmod 0600 "${redisConfVar}" "${redisConfRun}"
+          if [ ! -s ${redisConfVar} ]; then
+            echo 'include "${redisConfRun}"' > "${redisConfVar}"
+          fi
+          echo 'include "${redisConfStore}"' > "${redisConfRun}"
+          ${optionalString (conf.requirePassFile != null) ''
+            (echo -n "requirepass "
```
Comment (ju1m) [inline nixos/modules/services/databases/redis.nix:364]: Using `(` ... `)` forks a sub-shell, that's not necessary here, we can keep the original `{` ... `}`

**Review by ju1m — CHANGES_REQUESTED** [review]
@ajs124 , I'm not an advanced user of Redis, and I've never used anything in Redis needing it to rewrite its config file. AFAICS, this is only done when configuring a sentinel, but not otherwise. Meaning that only users advanced enough to use a sentinel will get a `/var/lib/${redisName name}/redis.conf` different than
a single include of `/run/${redisName name}/nixos.conf`.
If I'm right, then I'd say that a release note mentioning this new behavior is enough, and there's no need for an option.

**Review by ju1m — APPROVED** [review]

**Comment by ajs124** [comment]
> I've never used anything in Redis needing it to rewrite its config file

My understanding is that a user/client can just issue `CONFIG SET` and `CONFIG REWRITE` commands, as long as they have sufficient permissions. However, I assume barely any application does that.

File: nixos/modules/services/databases/redis.nix [hunk 3]
```diff
@@ -340,16 +347,24 @@ in {
...
+        in ''
+          touch "${redisConfVar}" "${redisConfRun}"
+          chown '${conf.user}' "${redisConfVar}" "${redisConfRun}"
+          chmod 0600 "${redisConfVar}" "${redisConfRun}"
+          if [ ! -s ${redisConfVar} ]; then
+            echo 'include "${redisConfRun}"' > "${redisConfVar}"
+          fi
+          echo 'include "${redisConfStore}"' > "${redisConfRun}"
+          ${optionalString (conf.requirePassFile != null) ''
+            {echo -n "requirepass "
```
Comment (andir) [inline nixos/modules/services/databases/redis.nix:364]: This doesn't look like valid bash anymore.

Comment (andir) [reply nixos/modules/services/databases/redis.nix:364]: nvm, this closing `}` just sits weird

Comment (ajs124) [reply nixos/modules/services/databases/redis.nix:364]: maybe
```diff
--- a/nixos/modules/services/databases/redis.nix
+++ b/nixos/modules/services/databases/redis.nix
@@ -361,8 +361,10 @@ in {
           fi
           echo 'include "${redisConfStore}"' > "${redisConfRun}"
           ${optionalString (conf.requirePassFile != null) ''
-            {echo -n "requirepass "
-            cat ${escapeShellArg conf.requirePassFile}} >> "${redisConfRun}"
+            {
+              echo -n "requirepass "
+              cat ${escapeShellArg conf.requirePassFile}
+            } >> "${redisConfRun}"
           ''}
         '');
         Type = "notify";
```
would be better?

**Pull request merged**